New Material Design calculator example.

MitchRudominer opened the pull request:
This example app is a Material design calculator that looks like the Android calculator.
It  is very simple but I'm willing to add more features in the future if anybody wants me too.
- Basic arithmetic operations only.
- Supports order of operations and gracefully handles infinity and NAN.
- Supports a DEL button using a stack of expressions
- I included some simple tests.

MitchRudominer committed:
New Material Design calculator example. (Attempt 2 after git fetch upstream).

abarth commented:
Travis is red.  We run the analyzer with a bunch of strict checks turned on.  You can reproduce those checks locally using `flutter analyze --flutter-repo`

abarth commented on examples/calculator/lib/calc_expression.dart:
```diff
@@ -0,0 +1,329 @@
+/// We need to represent pieces of numbers because the user may have only
+/// entered a partial expression so far.
+class ExpressionToken {
+  String stringRep;
```
final ?

MitchRudominer replied:
done

abarth commented on examples/calculator/lib/calc_expression.dart:
```diff
@@ -0,0 +1,329 @@
+
+  String toString() {
+    return stringRep;
+  }
```
We'd usually write this using the `=>` shorthand:

`String toString() => stringRep`

MitchRudominer replied:
done

abarth commented on examples/calculator/lib/calc_expression.dart:
```diff
@@ -0,0 +1,329 @@
+}
+
+// As the user taps different keys the current expression can be in one
+// of several states.
```
If you use `///` rather than `//`, these comments will be automatically picked up by the `dartdoc` tool as documentation.

MitchRudominer replied:
done

abarth commented on examples/calculator/lib/main.dart:
```diff
@@ -0,0 +1,258 @@
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+import 'calc_expression.dart';
```
Please add a blank line between `package` imports and path-relative imports.

See https://flutter.io/style-guide/ for the complete style guide.

MitchRudominer replied:
done

abarth commented on examples/calculator/lib/main.dart:
```diff
@@ -0,0 +1,258 @@
+  // As the user taps keys we update the current |_expression| and we also
+  // keep a stack of previous expressions so we can return to earlier states
+  // when the user hits the DEL key.
+  List<CalcExpression> _expressionStack = <CalcExpression>[];
```
final

MitchRudominer replied:
done

abarth commented:
This looks great.  The main issue is that it doesn't follow the style guide.  If you're interested in learning the style guide, you're welcome to read it and try to get the code to follow it, but if that doesn't sound like fun, I'm happy to do a style cleanup pass after it lands.

We should fix all the things that Travis checks though so we don't turn the tree red when landing.

Hixie commented on examples/calculator/test/calc_expression_test.dart:
```diff
@@ -0,0 +1,23 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
```
You should add this package to the list of packages to test in `.../travis/test.sh` before checkin.

MitchRudominer replied:
done

Hixie commented on examples/calculator/test/smoke_test.dart:
```diff
@@ -0,0 +1,34 @@
+    Finder display = find.widgetWithText(Flexible, '12');
+    expect(display, findsOneWidget);
+  });
+}
```
Some tests to add:
* 0.1 + 0.2 should output 0.3 (not 0.300000000 or similar)
* 1.0 / 10.0 should output 0.1
* check what happens with division by zero

MitchRudominer replied:
Done

MitchRudominer committed:
Responded to code review.

MitchRudominer commented:
ptal. I have fixed all of the Travis errors and linter warnings. I have responded to all of the code review comments. Regarding adherence to the style guide: Would you mind pointing out a few of the things that you want me to change and then I will change those things and anything else similar to those things that I find? Or if it is easier for you to do as you suggested and have you make all of the style changes then lets just do that.

Hixie commented:
We're happy to take care of it. I'll point out some things and if you decide it's not worth your time then that's fine. :-)

There's some merge conflicts which will need to be fixed though.

Hixie commented on examples/calculator/lib/calc_expression.dart:
```diff
@@ -0,0 +1,340 @@
+class ExpressionToken {
+  final String stringRep;
+
+  ExpressionToken(this.stringRep);
```
style nit: the constructor should come first.

MitchRudominer replied:
done (for all classes)

Hixie commented on examples/calculator/lib/calc_expression.dart:
```diff
@@ -0,0 +1,340 @@
+}
+
+/// A token that represents a number.
+class NumberToken extends ExpressionToken {
```
FWIW, I would probably make this a `NumberToken<T extends num>`, then have `FloatToken` inherit from `NumberToken<double>` and similar.

Hixie replied:
...although it looks like you have a number of subclasses that use both, so maybe not...

In general I would avoid using `num`, because it leads to polymorphic call sites, but performance isn't critical here so whatever.

MitchRudominer replied:
OK thanks for telling me. In this case I would like to leave it because otherwise it will make dealing with the results of computations quite a bit more laborious because I will have to keep track of whether or not they are pure int computations.

Hixie replied:
Yeah that seems totally reasonable in this case.

Hixie commented on examples/calculator/lib/calc_expression.dart:
```diff
@@ -0,0 +1,340 @@
+      : this.number = number,
+        super(stringRep);
+
+  NumberToken.fromNumber(num number) : this("$number", number);
```
style nit: we use single quotes rather than double quotes

MitchRudominer replied:
done

Hixie commented on examples/calculator/lib/calc_expression.dart:
```diff
@@ -0,0 +1,340 @@
+  final num number;
+
+  NumberToken(String stringRep, num number)
+      : this.number = number,
```
style nit: the `:` should be lined up under the `m` of `Number`. (This falls out of what would happen if the `)` was on this line.)

Hixie replied:
style nit:no need for the "this"

MitchRudominer replied:
changed

Hixie commented on examples/calculator/lib/calc_expression.dart:
```diff
@@ -0,0 +1,340 @@
+
+  static double _parse(String stringRep) {
+    String toParse = stringRep;
+    if (toParse.startsWith(".")) {
```
style nit: single-statement blocks don't get braces.

MitchRudominer replied:
done (but weird)

Hixie commented on examples/calculator/lib/calc_expression.dart:
```diff
@@ -0,0 +1,340 @@
+
+  // rounds |number| to 14 digits of precision. A double precision
+  // floating point number is guaranteed to have at least this many
+  // decimal digits of precision. If the inputs
```
this comment isn't complete

MitchRudominer replied:
done

Hixie commented on examples/calculator/lib/calc_expression.dart:
```diff
@@ -0,0 +1,340 @@
+  CalcExpression appendDigit(int digit) {
+    ExpressionState newState = ExpressionState.Number;
+    ExpressionToken newToken;
+    List<ExpressionToken> outList = _list.toList();
```
style nit: mark local variables that you don't change "final"

MitchRudominer replied:
done (here and elsewhere)

Hixie commented on examples/calculator/lib/calc_expression.dart:
```diff
@@ -0,0 +1,340 @@
+        break;
+      case ExpressionState.Number:
+        ExpressionToken last = outList.removeLast();
+        newToken = new IntToken(last.stringRep + "$digit");
```
style nit: we usually use the style `'${last.stringRep}$digit'` or similar

MitchRudominer replied:
done

Hixie commented on examples/calculator/lib/calc_expression.dart:
```diff
@@ -0,0 +1,340 @@
+    num currentValue = firstNumToken.number;
+    while (list.length > 0) {
+      bool isDivision = false;
+      OperationToken nextOpToken = list[0];
```
style nit: use `.first` instead of `[0]`.

MitchRudominer replied:
done

Hixie commented on examples/calculator/lib/calc_expression.dart:
```diff
@@ -0,0 +1,340 @@
+          isDivision = true;
+      }
+      // Remove the operation token.
+      list.removeAt(0);
```
`removeFirst`

MitchRudominer replied:
ack

Hixie commented on examples/calculator/lib/main.dart:
```diff
@@ -0,0 +1,261 @@
+// A calculator application.
+void main() {
+  runApp(new MaterialApp(
+      title: 'Calculator',
```
nit: indent by 2

Hixie commented on examples/calculator/lib/main.dart:
```diff
@@ -0,0 +1,261 @@
+  runApp(new MaterialApp(
+      title: 'Calculator',
+      theme: new ThemeData(primarySwatch: Colors.blue),
+      home: new Calculator()));
```
nit: if you have a newline after the `(`, there should be a newline before the `)`.

Hixie commented on examples/calculator/lib/main.dart:
```diff
@@ -0,0 +1,261 @@
+}
+
+class Calculator extends StatefulWidget {
+  Calculator({Key key}) : super(key: key);
```
nit: spaces just inside the braces around named optional arguments.

Hixie commented on examples/calculator/lib/main.dart:
```diff
@@ -0,0 +1,261 @@
+    _expression = resultExpression;
+  }
+
+  void onNumberTap(int n) {
```
nit: we name properties onFoo, functions that are assigned to those properties are handleFoo.

MitchRudominer replied:
done

Hixie commented on examples/calculator/lib/main.dart:
```diff
@@ -0,0 +1,261 @@
+class CalcDisplay extends StatelessWidget {
+  CalcDisplay(this._flex, this._contents);
+
+  int _flex;
```
fields of a Widget should all be final.

MitchRudominer replied:
done

Hixie commented on examples/calculator/lib/main.dart:
```diff
@@ -0,0 +1,261 @@
+    return new Flexible(
+        flex: _flex,
+        child: new Center(child: new Text(_contents,
+            style: new TextStyle(color: Colors.black, fontSize: 24.0))));
```
const TextStyle

MitchRudominer replied:
done

Hixie commented on travis/test.sh:
```diff
@@ -15,9 +15,10 @@ flutter analyze --flutter-repo
+(cd examples/flutter_gallery; flutter test)
 (cd examples/hello_world; flutter test)
 (cd examples/layers; flutter test)
-(cd examples/flutter_gallery; flutter test)
```
please leave flutter_gallery near the end, because (when the main test in there isn't disabled) it's one of the longest-running tests.

MitchRudominer replied:
done

Hixie commented on examples/calculator/lib/main.dart:
```diff
@@ -0,0 +1,261 @@
+}
+
+class MainKeyPad extends StatelessWidget {
+  final _CalculatorState calcState;
```
passing a State down to a subclass is a little sketchy. I think in your case it's guaranteed to be ok but in the general case it's possible for a state to get replaced but a child survive. Generally we'd have a chain of callbacks passed down at each level instead.

MitchRudominer replied:
I also didn't love the idea of passing the state object around like this.  But passing individual callbacks seems bad too here because I need a callback per button. So this solution seemed pragmatic.

Hixie commented on examples/calculator/lib/main.dart:
```diff
@@ -0,0 +1,261 @@
+}
+
+class CalcDisplay extends StatelessWidget {
+  CalcDisplay(this._flex, this._contents);
```
for widely-used widgets, we have the convention of always exposing a Key argument and having all the arguments be named. But you don't have to do that here since they are all internal widgets.

MitchRudominer replied:
ack

MitchRudominer committed:
Second round of code review.

MitchRudominer commented:
ptal. I have addressed all comments except two having to do with formatting (indentation etc.) I really don't want to fight with the auto-code formatter. Every time I try to make the changes you suggest at some point later I will run the code formatter and it reverses the chang.

Hixie commented:
Thanks! I'll take a look.

Yeah, the Dart formatter isn't compatible with out style guide. We'll probably have our own formatter in due course.

Hixie commented:
![LGTM](http://junkyard.damowmow.com/563)

Hixie commented:
Would you like me to check this in for you?
